DataScience/Analytics/prototype/landRegistry.py

```diff
@@ -0,0 +1,846 @@
+            parsed['Postcode'] = pcode
+
+        if parsed.get('Postcode', None) is not None:
+            # check that there is space, if not then add
+            if ' ' not in parsed['Postcode']:
+                inc = parsed['Postcode'][-3:]
+                out = parsed['Postcode'].replace(inc, '')
+                parsed['Postcode'] = out + ' ' + inc
+
+            # change to all capitals
+            parsed['Postcode'] = parsed['Postcode'].upper()
+
+        # if Hackney etc. in StreetName then remove and move to locality if town name contains London
+        # todo: probabilistic parser should see more cases with london localities, parsed incorrectly at the mo
+        if parsed.get('StreetName', None) is not None and parsed.get('TownName', None) is not None:
+            if 'LONDON' in parsed['TownName']:
+                # todo: should move to a file rather than have inside the code and get a complete list from AB
+                locs = ['HACKNEY', 'ISLINGTON', 'STRATFORD', 'EAST HAM', 'WOOD GREEN', 'FINCLEY', 'HORNSEY', 'HENDON',
+                        'TOTTENHAM', 'BLACKHEATH', 'BAYSWATER', 'CHISWICK VILLAGE', 'CHISWICK', 'COLINDALE', 'LEWISHAM',
+                        'FOREST HILL',  'NORBURY', 'MANOR PARK', 'PLAISTOW', 'ABBEY WOOD', 'SOUTH NORWOOD', 'CHARLTON',
+                        'MOTTINGHAM', 'NEW ELTHAM', 'BATTERSEA', 'PUTNEY', 'TOOTING', 'RAYNES PARK', 'MORTLAKE',
+                        'WEST KENSINGTON', 'KENSINGTON', 'ACTON', 'HAMMERSMITH', 'HANWELL', 'NEW SOUTHGATE',
+                        'GREEN LANES', 'STREATHAM HILL', 'CATFORD', 'LEWISHAM', 'BALHAM', 'OLYMPIC PARK', 'CHINGFORD',
+                        'STREATHAM', 'LEYTONSTONE', 'BROCKLEY', 'SOUTH WALTHAMSTOW', 'WALTHAMSTOW', 'MAIDA VALE',
+                        'HOLLAND PARK', 'FULHAM', 'PARK ROYAL', 'LEYTON', 'TULSE HILL', 'SILVERTOWN', 'WOODFORD',
+                        'ROYAL VICTORIA DOCK', 'CROUCH END', 'EDMONTON', 'PLUMSTEAD', 'ELTHAM', 'EAST DULWICH',
+                        'MUSWELL HILL', 'EALING', 'WANSTEAD', 'WIMBLEDON', 'UPPER NORWOOD', 'CAMBERWELL', 'SYDENHAM',
+                        'SOUTHFIELDS', 'COLLIERS WOOD', 'THAMESMEAD', 'WILLESDEN', 'HAMPSTEAD', 'KILBURN',
+                        'KENTISH TOWN', 'HARLESDEN', 'FULHAM', 'WEST DULWICH', 'LONDON', 'PALMERS GREEN', 'MARYLEBONE',
+                        'CLAPHAM', 'WANDSWORTH', 'WOOLWICH', 'BELLINGHAM', 'GREENWICH', 'NEW CROSS', 'KIDBROOKE',
+                        'HOMERTON']
+
+                for loc in locs:
+                    if parsed['StreetName'].strip().endswith(loc):
+                        parsed['Locality'] = loc
+                        # take the last part out, so that e.g. CHINGFORD AVENUE CHINGFORD is correctly processed
+                        # need to be careful with e.g.  WESTERN GATEWAY ROYAL VICTORIA DOCK
+                        parsed['StreetName'] = ' '.join(parsed['StreetName'].strip().split(' ')[:-len(loc.split(' '))])
```

Comment:
It admit that my solution may be overly complicated, however, I am not only cutting out the last word but as many words as the "loc" variable may refer to, which can be any number (usually less than 3 though). Hence the ":-len(loc.split(' '))" part, as I don't know a priori how many parts from the end I need to remove.
I should probably use rsplit like you suggest and change the line to something like "text.strip().rsplit(' ', len(loc.split(' '))[0]". That's probably slightly better than the .join, even if the slicing still requires "len".